File: core/src/test/java/javabot/operations/JavadocOperationTest.java
package javabot.operations;

import com.antwerkz.sofia.Sofia;
import com.jayway.awaitility.Duration;
import javabot.BaseMessagingTest;
import javabot.Messages;
import javabot.dao.ApiDao;
import javabot.dao.EventDao;
import javabot.javadoc.JavadocApi;
import javabot.model.ApiEvent;
import org.testng.annotations.BeforeTest;
import org.testng.annotations.Test;

import javax.inject.Inject;
import java.net.MalformedURLException;
import java.util.concurrent.TimeUnit;

@Test//(dependsOnMethods = {"jdk"})
public class JavadocOperationTest extends BaseMessagingTest {
    @Inject
    private ApiDao apiDao;
    @Inject
    private EventDao eventDao;
    @Inject
    private Messages messages;

    @BeforeTest
    public void getBot() {
        getJavabot();
    }

    public void jdk() throws MalformedURLException {
        JavadocApi api = apiDao.find("JDK");
        if (api == null) {
            ApiEvent event = new ApiEvent(getTestUser().getNick(), "JDK", "http://docs.oracle.com/javase/8/docs/api",
                                          null);
            eventDao.save(event);
            waitForEvent(event, "adding JDK", new Duration(30, TimeUnit.MINUTES));
        }
        messages.get();

    }

    public void constructors() throws MalformedURLException {
        jdk();
        scanForResponse("~javadoc java.lang.String(char[])", "java/lang/String.html");
        scanForResponse("~javadoc String(char[])", "java/lang/String.html#String-char[]-");
    }

    public void methods() throws MalformedURLException {
        jdk();
        scanForResponse("~javadoc String.split(String)", "java/lang/String.html#split-java.lang.String-");
        scanForResponse("~javadoc -jdk String.split(String)", "java/lang/String.html#split-java.lang.String-");
        scanForResponse("~javadoc String.split(java.lang.String)", "java/lang/String.html#split-java.lang.String-");
        scanForResponse("~javadoc String.join(*)", "java/lang/String.html#join-");
        scanForResponse(String.format("%s %s", getJavabot().getNick(), "javadoc String.split(*)"),
                        "java/lang/String.html#split-java.lang.String-");
    }

    public void nestedClasses() throws MalformedURLException {
        jdk();
        scanForResponse("~javadoc Map.Entry", "java/util/Map.Entry.html");
    }

    public void format() throws MalformedURLException {
        jdk();
        scanForResponse("~javadoc String.format(*)", "java/lang/String.html#format-java.util.Locale-java.lang.String-java.lang.Object...");
    }

    @Test
    public void doFinal() throws MalformedURLException {
        jdk();
        scanForResponse("~javadoc String.valueOf(*)", Sofia.tooManyResults(getTestUser().getNick()));
    }

    public void fields() throws MalformedURLException {
        jdk();
        scanForResponse("~javadoc System.in", "java/lang/System.html#in");
        scanForResponse("~javadoc Integer.MAX_VALUE", "java/lang/Integer.html#MAX_VALUE");
        scanForResponse("~javadoc -jdk System.in", "java/lang/System.html#in");
    }

    public void inherited() throws MalformedURLException {
        jdk();
        scanForResponse("~javadoc ArrayList.listIterator(*)", "java/util/ArrayList.html#listIterator-int-");
    }
}

File: src/test/kotlin/javabot/qtest/operations/JavadocOperationTest.kt
package javabot.qtest.operations

import com.antwerkz.sofia.Sofia
import io.quarkus.test.junit.QuarkusTest
import javabot.BaseTest
import javabot.dao.JavadocClassDao
import javabot.operations.JavadocOperation
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

@QuarkusTest
class JavadocOperationTest : BaseTest() {
    companion object {
        val STRING_URL = "java.base/java/lang/String.html#"
    }

    private val operation: JavadocOperation by lazy {
        injector.getInstance(JavadocOperation::class.java)
    }
    private val javadocClassDao: JavadocClassDao by lazy {
        injector.getInstance(JavadocClassDao::class.java)
    }

    @BeforeEach
    fun jdk() {
        loadApi("JDK", version = "11")
    }

    @Test
    fun constructors() {
        scanForResponse(
            operation.handleMessage(message("~javadoc String(char[])")),
            "$STRING_URL<init>(char[])",
        )
        scanForResponse(
            operation.handleMessage(message("~javadoc java.lang.String(char[])")),
            "$STRING_URL<init>(char[])",
        )
    }

    @Test
    fun methods() {
        scanForResponse(
            operation.handleMessage(message("~javadoc String.split(String)")),
            "${STRING_URL}split(java.lang.String)",
        )
        scanForResponse(
            operation.handleMessage(message("~javadoc -jdk String.split(String)")),
            "${STRING_URL}split(java.lang.String)",
        )
        scanForResponse(
            operation.handleMessage(message("~javadoc String.split(java.lang.String)")),
            "${STRING_URL}split(java.lang.String)",
        )
        scanForResponse(
            operation.handleMessage(message("~javadoc String.join(*)")),
            "${STRING_URL}join(",
        )
        scanForResponse(
            operation.handleMessage(message("~javadoc String.split(*)")),
            "${STRING_URL}split(java.lang.String)",
        )
    }

    @Test
    fun nestedClasses() {
        scanForResponse(
            operation.handleMessage(message("~javadoc Map.Entry")),
            "java/util/Map.Entry.html",
        )
    }

    @Test
    fun format() {
        scanForResponse(
            operation.handleMessage(message("~javadoc String.format(*)")),
            "java/lang/String.html#format(java.util.Locale,java.lang.String,java.lang.Object...)",
        )
    }

    @Test
    fun doFinal() {
        scanForResponse(
            operation.handleMessage(message("~javadoc String.valueOf(*)")),
            Sofia.tooManyResults(TEST_USER.nick),
        )
    }

    @Test
    fun fields() {
        scanForResponse(
            operation.handleMessage(message("~javadoc System.in")),
            "java/lang/System.html#in",
        )
        scanForResponse(
            operation.handleMessage(message("~javadoc Integer.MAX_VALUE")),
            "java/lang/Integer.html#MAX_VALUE",
        )
        scanForResponse(
            operation.handleMessage(message("~javadoc -jdk System.in")),
            "java/lang/System.html#in",
        )
    }

    @Test
    fun inherited() {
        scanForResponse(
            operation.handleMessage(message("~javadoc ArrayList.listIterator(*)")),
            "java/util/ArrayList.html#listIterator(int)",
        )
    }

    @Test
    fun packagePrivate() {
        scanForResponse(
            operation.handleMessage(message("~javadoc ASCII)")),
            "I have no documentation for ASCII",
        )
    }
}
